src/surface_finder.cpp

```diff
@@ -0,0 +1,316 @@
+#include "pcl/filters/extract_indices.h"
+#include "pcl/point_cloud.h"
+#include "pcl/point_types.h"
+#include "ros/ros.h"
+
+typedef pcl::PointXYZRGB PointC;
+typedef pcl::PointCloud<pcl::PointXYZRGB> PointCloudC;
+
+namespace {
+/**
+ * This is a helper function for calculating angle between given plane and z
+ * axis
+ */
+double calculateAngle(const double& a, const double& b, const double& c) {
+  double denominator = sqrt(pow(a, 2) + pow(b, 2) + pow(c, 2));
+  double nominator = fabs(c);
+  double tmpRes = nominator / denominator;
+  return acos(tmpRes);
+}
+
+/**
+ * This function finds the coefficients of a plane equation, given three points
+ */
+void planeEquation(const std::vector<PointC>& pts, double* a, double* b,
+                   double* c, double* d) {
+  *a = (pts[1].y - pts[0].y) * (pts[2].z - pts[0].z) -
+       (pts[2].y - pts[0].y) * (pts[1].z - pts[0].z);
+  *b = (pts[1].z - pts[0].z) * (pts[2].x - pts[0].x) -
+       (pts[2].z - pts[0].z) * (pts[1].x - pts[0].x);
+  *c = (pts[1].x - pts[0].x) * (pts[2].y - pts[0].y) -
+       (pts[2].x - pts[0].x) * (pts[1].y - pts[0].y);
+  *d = -1 * (*a) * pts[0].x - (*b) * pts[0].y - (*c) * pts[0].z;
+}
+
+/**
+ * This function gives a vector of points within certain distance to a plane
+ * specified by the z_val
+ */
+void filterIndices(const double& dist_limit,
+                   const std::map<double, std::vector<int> >& sortedIndices,
+                   const double& z_val, std::vector<int>* output_pts) {
+  std::map<double, std::vector<int> >::const_iterator iter =
+      sortedIndices.find(z_val);
+  if (iter == sortedIndices.end()) {
+    ROS_INFO("Error: sampled point with height %f not on the plane", z_val);
+    return;
+  }
+
+  std::map<double, std::vector<int> >::const_iterator curr_iter = iter;
+  iter++;
+
+  // Find the indices below z_val
+  while (curr_iter != sortedIndices.begin() &&
+         (z_val - curr_iter->first) <= dist_limit) {
+    for (size_t i = 0; i < curr_iter->second.size(); i++) {
+      output_pts->push_back(curr_iter->second[i]);
+    }
+    curr_iter--;
+  }
+  if (curr_iter == sortedIndices.begin() &&
+      (z_val - curr_iter->first) <= dist_limit) {
+    for (size_t i = 0; i < curr_iter->second.size(); i++) {
+      output_pts->push_back(curr_iter->second[i]);
+    }
+  }
+
+  // Find the indices above z_val
+  curr_iter = iter;
+  while (curr_iter != sortedIndices.end() &&
+         (curr_iter->first - z_val) <= dist_limit) {
+    for (size_t i = 0; i < curr_iter->second.size(); i++) {
+      output_pts->push_back(curr_iter->second[i]);
+    }
+    curr_iter++;
+  }
+}
+
+/**
+ * This function determines if given two planes are similar
+ */
+bool isSimilar(const double& dist, const pcl::ModelCoefficients& plane1,
+               const pcl::ModelCoefficients& plane2) {
+  double z1 = -1.0 * plane1.values[3] / plane1.values[2];
+  double z2 = -1.0 * plane2.values[3] / plane2.values[2];
+  return fabs(z1 - z2) < 10 * dist;
```

Comment:
Why is there a 10 here?

Reply:
Changed to 2 to prevent intersection

Reply:
Previously, with 10 as the multiplier, the surface after refinement is still considered the same as the original surface by isSimilar function.

Now, with 2 as the multiplier, the refined surface is considered "different" from the original surface, because their z-axis intercept are two far away. The reason behind it is that 2 makes isSimilar has a more strict requirement for similar surface.

This should not affect the performance of the algorithm if the client has good parameter for magin_above_surface, with correct minimum number of points to be a surface.